nasa/ow_simulator
# Oceanwater 1009 implement telemetry system faults status

## Linked Issues:
| EPIC ⚡ | [OCEANWATER-934](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-934) |
| :----------- | :----------- |
| Jira Ticket 🎟️ | [OCEANWATER-1009](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-1009) |
| Github :octocat:  | [OCEANWATER-1009_implement_telemetry_system_faults_status](https://github.com/nasa/ow_simulator/tree/OCEANWATER-1009_implement_telemetry_system_faults_status) |

First of a group of Telemetry Unification tickets (listed below) targeting fault related messages in accordance with OceanWATERS/OWLAT Unification Plan. Mostly artifact renaming and updating message definitions.

**This issue has a corresponding and required branch on [ow_autonomy](https://github.com/nasa/ow_autonomy/tree/OCEANWATER-1009_implement_telemetry_system_faults_status).**

Related to the following tickets:
* [OCEANWATER-1010](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-1010)
* [OCEANWATER-1011](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-1011)
* [OCEANWATER-1013](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-1013)
* [OCEANWATER-1014](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-1014)

## Summary of Changes
The largest change, which will necessarily be present in the grouped tickets above, is the creation of the `owl_msgs` package. The new `owl_msgs` package houses telemetry messages used by the fault detection and injection packages, according to the OWLAT-OceanWATERS Telemetry Unification plan. The changes in these tickets affect the [ow_autonomy]() and [ow_sim](https://github.com/nasa/ow_simulator/tree/OCEANWATER-1009_implement_telemetry_system_faults_status) branches.

* Created the `owl_msgs` package.
* Updated `ow_faults_detection` to depend on `owl_msgs`

Specific to this ticket:
* Renamed the `SystemFaults` message to `SystemFaultsStatus` and updated all references in source and documentation.
* Moved `SystemFaultsStatus.msg` to the new `owl_msgs` package.
* Removed the `/faults/` topic prefix from the `system_faults_status` topic.
* Updated `SystemFaultsStatus` message semantics both in naming and to include more aggregate error entries.
* Made Plexil-OwInterface error consistent with updated message semantics.
* Renamed some variables to be consistent with message rename.
* Removed now redundant bitsets in favor of using msg constants defined in SystemFaultsStatus.msg

## Test
* Build according to [User Documentation](https://github.com/nasa/ow_simulator/wiki/Using-OceanWATERS).
* Run two terminals sourced with the project's `setup.bash`.
* In one terminal, run the project:
  * `roslaunch ow atacama_y1a.launch`
* In the second terminal, verify the fault messages are in the `owl_msgs` package and its contents:
  * `rosmsg package owl_msgs`
  * `rosmsg show owl_msgs/SystemFaultsStatus`
* In the second terminal Subscribe to the system faults status topic
  * `rostopic echo /system_faults_status`
* In RQT go to the Dynamic Reconfigure tab and select faults.
* Toggle any of the faults (ie: arm_faults) and observe the rostopic change with an appropriate value

## Comments/Issues
* It appears that the only fault error codes are implemented in the latest development branch `noetic-devel`
